elastic/kibana
# [Cases] Custom fields MVP

## Summary

This PR implements an MPV for custom fields in Cases. Specifically:

- Users can add, delete, or edit custom fields from the configuration page
- Users can set custom fields when creating a case
- Users can set custom fields when editing a case
- Only the `Text` custom field type is supported. The `Toggle` is implemented to test the framework.
- Users with a basic license can enter the configuration page and set custom fields
- The configuration page header changed to "Settings"
- The "Edit external connection" button changed to "Settings"
- APIs:
    - Users cannot post custom fields with duplicate keys
    - Users cannot change the type of the custom field
    - Users cannot add custom fields that are not configured
    - Required fields should be present
    - If some of the custom fields are omitted from the request the backend will fill them with `null` values 
- Limits:
    - Maximum 10 custom fields configured
    - Maximum 50 chars in custom field labels
    - Maximum 160 chars for the value of the `Text` custom field type
    - Users cannot change the type of a custom field
    - The key of the custom fields should contain only lowercase letters (a-z), numbers (0-9), '_', and '-'
    - Required fields need a value

## Testing
- Cases created before custom fields are working as expected (create some on `main` before switching to the feature branch)
- Environments without configuration are working as expected (clear ES data)
- Environments with existing configurations are working as expected
- Try to create cases with custom fields (required & optional) and delete some of them
- Try to create cases with custom fields (required & optional) and then configure new custom fields. Existing cases with old custom fields should work as expected
- User actions are working as expected especially with the combinations described in the previous two bullets
- Users with a basic license can configure custom fields but cannot configure connectors
- Users with a gold+ license can configure custom fields and connectors
- Users with read access cannot configure or edit custom fields

Resolves: https://github.com/elastic/kibana/issues/160236

PRs:

- https://github.com/elastic/kibana/pull/165671
- https://github.com/elastic/kibana/pull/166353
- https://github.com/elastic/kibana/pull/166439
- https://github.com/elastic/kibana/pull/166483
- https://github.com/elastic/kibana/pull/166815
- https://github.com/elastic/kibana/pull/166940
- https://github.com/elastic/kibana/pull/166962
- https://github.com/elastic/kibana/pull/166969
- https://github.com/elastic/kibana/pull/166975
- https://github.com/elastic/kibana/pull/167029
- https://github.com/elastic/kibana/pull/167047
- https://github.com/elastic/kibana/pull/167105
- https://github.com/elastic/kibana/pull/167131
- https://github.com/elastic/kibana/pull/167144
- https://github.com/elastic/kibana/pull/167166
- https://github.com/elastic/kibana/pull/167167
- https://github.com/elastic/kibana/pull/167310
- https://github.com/elastic/kibana/pull/167386
- https://github.com/elastic/kibana/pull/167481
- https://github.com/elastic/kibana/pull/167495
- https://github.com/elastic/kibana/pull/167398
- https://github.com/elastic/kibana/pull/167472

### Checklist

Delete any items that are not applicable to this PR.


### For maintainers


## Release notes
Allow users to configure custom fields in Cases. Supported field types: Text and Toggle. Coming soon: List, Number, Date, IP, Email, etc.
